### PR TITLE
fix(N-06): used namespaced storage

### PR DESCRIPTION
**Pull request by DinshawKothari**
Ticket: USDS-128

### CODE REVIEW

**Review by sachushaji — COMMENTED** [review]
Small lint issues

**Comment by DinshawKothari** [comment]
> Small lint issues

Done.

File: contracts/USDS.sol [hunk 1]
```diff
@@ -59,6 +56,20 @@ contract USDS is
...
     }
 
+    // --- Namespaced Storage ---
+    struct USDSStorage {
+        AggregatorV3Interface proofOfReserveFeed;
+        uint256 acceptableProofOfReserveTimeDelay;
+        uint256 mintCapPerTransaction;
+    }
+
+    function _getUSDSStorage() private pure returns (USDSStorage storage s) {
```
Comment (sachushaji) [inline contracts/USDS.sol]: Can we move this towards the end, according to the style guide

File: contracts/USDS.sol [hunk 2]
```diff
@@ -59,6 +56,20 @@ contract USDS is
...
+        bytes32 slot = USDSStorageLocation;
+        assembly {
+            s.slot := slot
+        }
```
Comment (sachushaji) [inline contracts/USDS.sol]:
```suggestion
        assembly {
            s.slot := USDSStorageLocation
        }
```

File: contracts/USDS.sol (hunk 1 above)
Comment (sachushaji) [reply contracts/USDS.sol]: Add natspec comments also to the new functions

File: contracts/USDS.sol [hunk 3]
```diff
@@ -20,7 +20,7 @@ contract USDS is
 {
     using SafeERC20 for IERC20;
 
-    AggregatorV3Interface private proofOfReserveFeed;
+    bytes32 private constant USDSStorageLocation = keccak256("bitgo.contracts.storage.USDS");
```
Comment (sachushaji) [inline contracts/USDS.sol]:
```suggestion
    bytes32 private constant USDSStorageLocation = keccak256("contract.storage.USDS");
```

File: contracts/USDS.sol (hunk 2 above)
Comment (DinshawKothari) [reply contracts/USDS.sol]: This is not possible as -
Only direct number constants and references to such constants are supported by inline assembly.

File: contracts/USDS.sol [hunk 4]
```diff
@@ -428,4 +453,17 @@ contract USDS is
...
+     * @dev Fetches the namespaced storage structure for the USDS contract.
+     * This function uses EIP-7201-style namespaced storage to ensure compatibility
+     * and extensibility for upgradeable contracts.
+     * @return $ The `USDSStorage` struct containing storage variables specific to the USDS contract.
+     */
+    function _getUSDSStorage() private pure returns (USDSStorage storage $) {
+        bytes32 slot = USDSStorageLocation;
+        assembly {
+            $.slot := slot
+        }
```
Comment (sachushaji) [inline contracts/USDS.sol]:
```suggestion
        assembly {
            $.slot := USDSStorageLocation
        }
```

Comment (DinshawKothari) [reply contracts/USDS.sol]: https://github.com/BitGo/USDS/pull/41#discussion_r1872694890

Comment (sachushaji) [reply contracts/USDS.sol]: Then lets calculate that and use a number constant which represents the slot

**Comment by sachushaji** [comment]
May need to move the ROLE hash also to the namespaced storage

File: contracts/USDS.sol (hunk 4 above)
Comment (DinshawKothari) [reply contracts/USDS.sol]: sure.

**Comment by DinshawKothari** [comment]
> May need to move the ROLE hash also to the namespaced storage

I do not think we need to change this, as they do not fall under the category of storage slots. They do not represent storage variables. Instead, they are constants that store the keccak256 hash of role identifiers. They are just immutable values in the bytecode. Also, they have not mentioned any change for role hash in audit report.

**Review by Phani024 — APPROVED** [review]

**Pull request merged**